### ISSUE TITLE
ops/triton/linear.py kernel_fwd computes its bounds mask and never passes it to the store

### ISSUE THREAD

**Issue by truong-v** [Dao-AILab/flash-attention#2789]
## Summary

`kernel_fwd` builds exactly the bounds mask its output store needs, on the line directly above the store, and then calls `tl.store(C, acc)` without it. The host grid rounds up with `cdiv`, so whenever `M` or `N` is not a multiple of the autotuned block sizes the last tiles write their full `BLOCK_M × BLOCK_N` extent: past the end of the `(M, N)` output allocation, and — because an N-tail overshoot lands on the next row of the same tensor — onto cells owned by other tiles, so the in-bounds output is also wrong, nondeterministically. The sibling `kernel_bwd` in the same file has the identical three lines and stores with `mask=mask`.

This reaches users through `triton_linear_act` / `FusedDenseSqreluDense` (`flash_attn/ops/triton/mlp.py`). Two scoping notes: tile-aligned shapes (powers of two) hide the bug completely, which is why tests and benchmarks never see it; and the module imports `triton.ops.matmul_perf_model`, which no longer exists in triton ≥ 3.0 — so the affected code is usable (and broken) only on triton 2.x builds (torch ≤ 2.3).

## Reproduction

```python
import torch
from flash_attn.ops.triton.linear import triton_linear_act

torch.manual_seed(0)
M, N, K = 513, 1000, 256
x = torch.randn(M, K, device="cuda", dtype=torch.float16)
w = torch.randn(N, K, device="cuda", dtype=torch.float16)
ref = (x.float() @ w.float().t()).half()
for _ in range(3):
    out = triton_linear_act(x, w)
    print(int((~torch.isclose(out, ref, atol=1e-1, rtol=1e-1)).sum()), "wrong cells of", M * N)
```

Output at a369df7 (L40S, torch 2.3.1 / triton 2.3.1):

```
[M=513 N=1000 K=256] wrong output cells vs x @ w.T over 3 runs: [6472, 6113, 10358] of 513000
[M=512 N=1024 K=256 (tile-aligned control)] wrong output cells: [0, 0, 0]
```

The varying counts across identical runs are the tile race on the in-bounds cells; under `compute-sanitizer --tool memcheck` (with `PYTORCH_NO_CUDA_MEMORY_CACHING=1`) a single small call at M=5, N=17, K=9 reports invalid global writes attributed to the store at `linear.py:255` and then ends in an unspecified launch failure, and passing the mask that is already computed there gives 0 wrong cells, deterministic, and 0 sanitizer errors.

## Environment

- flash-attention commit a369df7 (main); flash_attn 2.7.4.post1 wheel for the compiled extension, checkout sources for the Triton module
- NVIDIA L40S (sm_89), driver CUDA 12.4
- torch 2.3.1+cu121, triton 2.3.1, Python 3.10
